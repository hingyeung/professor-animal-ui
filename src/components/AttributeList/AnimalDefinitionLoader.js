import React from 'react';

const AnimalDefinitionLoader = (props) => {
  function onChange(event) {
    const fileToOpen = event.target.files[0];

    if (fileToOpen && fileToOpen.type === 'application/json') {
      let fileReader = new FileReader();

      fileReader.onload = function(e) {
        let jsonStr = fileReader.result;
        props.onFileLoaded(jsonStr);
      };

      fileReader.readAsText(fileToOpen);
    }
  }

  return (
    <div>
      <form>
        <input type="file" className="form-control" onChange={onChange}/>
      </form>
    </div>
  );
};

export default AnimalDefinitionLoader;