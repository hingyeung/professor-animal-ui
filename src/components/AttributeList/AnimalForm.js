import React, {Component} from 'react';
import AttributeGroup from "./AttributeGroup";

class AnimalForm extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    let attributeGroupContent = [];

    const currentAnimal = this.props.animal;
    Object.keys(currentAnimal.attributeMap).forEach(attributeType => {
      attributeGroupContent.push(
        <AttributeGroup key={ attributeType + currentAnimal.id }
                        attributeType={ attributeType }
                        attributes={ currentAnimal.attributeMap[attributeType] }/>
      );
    });

    return (
      <form onSubmit={ this.props.onFormSubmit }>
        <div>
          <button className="btn btn-primary">Save</button>
        </div>
        <div className="form-group row">
          <label htmlFor="animal-name" className="col-sm-2 col-form-label">Animal name</label>
          <div className="col-sm-6">
            <input id="animal-name" className="form-control" type="text" name="animalName"
                   onChange={ (e) => this.props.updateAnimalName(e.target.value) } value={ currentAnimal.name }/>
          </div>
        </div>
        { attributeGroupContent }
        <div>
          <button className="btn btn-primary">Save</button>
        </div>
      </form>
    )
  }
}

export default AnimalForm;