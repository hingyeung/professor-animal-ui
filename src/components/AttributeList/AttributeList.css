.attribute-list-container {
    /*overflow: scroll;*/
    /*height: 100%;*/
}

.attribute-list-container table thead th {
    border: none;
    text-transform: capitalize;
}

.left-container, .right-container {
    overflow-y: scroll;
    position: absolute;
    top: 0;
    bottom: 0;
}

.left-container {
    left: 0;
    background-color: #80bdff;
}

.right-container {
    right: 0;
    background-color: thistle;
}