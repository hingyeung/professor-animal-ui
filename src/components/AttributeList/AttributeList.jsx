import './AttributeList.css';

import React, {Component} from 'react';
import {guid} from 'services/GUID';
import update from 'immutability-helper';
import {
  BrowserRouter as Router,
  Route,
  Switch,
  Link
} from 'react-router-dom';
import Greeting from 'components/Greeting';
import AnimalDefinitionLoader from 'components/AnimalDefinitionLoader';
import {asyncImport} from 'services/AsyncImport';
import AnimalDefinition from "models/AnimalDefinition";
import AttributeDefinitionReaderService from 'services/AttributeDefinitionReaderService';
import PropTypes from 'prop-types';

const AnimalList = asyncImport('AnimalList');
const AnimalForm = asyncImport('AnimalForm');

class AttributeList extends Component {
  static HOME = "/";
  static NEW = "/new";
  static LOADFILE = "/load";
  static ANIMAL_UPDATED = "/updated";

  constructor(props) {
    super(props);

    this.state = {
      // The attribute definition loaded from file as the base template for new animal form.
      // e.g. { physical: [{_name: "fins", _value: true}]}
      attributeDefinition: {},
      // This is the state of the current animal definition. This should be updated when an attribute of
      // an animal is changed, a new attribute is added to an animal, or when a new animal is added.
      // e.g. {id: {name: "Lion", attributeMap: {physical: {tail: true, legs: true}}}}
      animalDefinition: {}
    };

    this.renderAttributeGroupsForThisAnimal = this.renderAttributeGroupsForThisAnimal.bind(this);
    this.onFormSubmit = this.onFormSubmit.bind(this);
    this.renderAttributeGroupsForNewAnimal = this.renderAttributeGroupsForNewAnimal.bind(this);
    this.populateAnimalForNewAnimalForm = this.populateAnimalForNewAnimalForm.bind(this);
    this.renderAnimalList = this.renderAnimalList.bind(this);
    this.onAnimalDefinitionLoaded = this.onAnimalDefinitionLoaded.bind(this);
    this.renderAnimalDefinitionLoader = this.renderAnimalDefinitionLoader.bind(this);
  }

  // When the animal definition is loaded from file, each animal definition only contains attributes that it has,
  // but not the attributes that it doesn't have. This is done to save storage space. However, the AnimalForm
  // that should allow user to modify all attributes, regardless whether the attributes are currently selected
  // for an animal for not.
  // This function hydrates the animal definition loaded from file to make sure all animals have all attributes
  // selectable on the form (added attributes are default to false).
  static addUnusedAttributeToAnimals(oldAnimalDefinition, attributeDefinition) {
    // let attributeDefinition = props.attributeDefinition;
    // must do deep cloning for hydratedAnimalDefinition would just be a
    // reference to oldAnimalDefinition
    let hydratedAnimalDefinition = JSON.parse(JSON.stringify(oldAnimalDefinition));

    if (attributeDefinition) {
      Object.keys(attributeDefinition).forEach(attributeType => {
        Object.keys(attributeDefinition[attributeType]).forEach(attributeName => {
          Object.keys(hydratedAnimalDefinition).forEach(animalId => {
            if (!oldAnimalDefinition[animalId]['attributeMap'][attributeType][attributeName]) {
              hydratedAnimalDefinition[animalId]['attributeMap'][attributeType][attributeName] =
                // default value is false from attributeDefinition
                attributeDefinition[attributeType][attributeName];
            }
          });
        })
      });
    }

    return hydratedAnimalDefinition;
  }

  onFormSubmit(animal, routeHistory) {
    this.setState({
      animalDefinition: update(this.state.animalDefinition, this.updateObjectForAnimal(animal))
    }, routeHistory.push(AttributeList.ANIMAL_UPDATED));
  }

  updateObjectForAnimal(animal) {
    return {
      [animal.id]: {
        $set: animal
      }
    }
  }

  // populate animal object with attributeDefinition for new animal form
  populateAnimalForNewAnimalForm() {
    return {
      name: undefined,
      id: guid(),
      attributeMap: this.state.attributeDefinition
    }
  };

  renderAttributeGroupsForNewAnimal(routeProps) {
    const routeHistory = routeProps.history;
    return <AnimalForm
      animal={ this.populateAnimalForNewAnimalForm() } // new animal
      routeHistory={ routeHistory }
      onFormSubmit={ this.onFormSubmit }
    />
  }

  renderAttributeGroupsForThisAnimal(routeProps) {
    const animalId = routeProps.match.params.id;
    const routeHistory = routeProps.history;
    const animal = this.state.animalDefinition[animalId];

    if (!animal) {
      // cannot push anything to routeHistory as it will call
      // setState() eventually and it is not allowed during render().
      // routeHistory.push(AttributeList.LOADFILE);
      return <Greeting/>;
    }
    return <AnimalForm
      animal={ this.state.animalDefinition[animalId] }
      routeHistory={ routeHistory }
      onFormSubmit={ this.onFormSubmit }/>
  }

  onExport(e) {
    e.preventDefault();
    this.props.onExport(this.state.animalDefinition);
  }

  renderAnimalList(routeProps) {
    const activeAnimalId = routeProps.location.state && routeProps.location.state.activeAnimalId ?
      routeProps.location.state.activeAnimalId :
      undefined;
    return <AnimalList activeAnimalId={ activeAnimalId } animals={ this.state.animalDefinition }/>
  }

  onAnimalDefinitionLoaded(jsonStrLoadedFromFile) {
    this.setState({
      // AttributeList.addUnusedAttributeToAnimals(props.animalDefinition, props.attributeDefinition)
      animalDefinition: AttributeList.addUnusedAttributeToAnimals(
        AnimalDefinition.convertFromFileModelToAppModel(JSON.parse(jsonStrLoadedFromFile)),
        this.state.attributeDefinition)
    })
  }

  renderAnimalDefinitionLoader(routeProps) {
    return <AnimalDefinitionLoader onFileLoaded={this.onAnimalDefinitionLoaded}/>
  }

  componentDidMount() {
    let attributeDefinitionReaderService = new AttributeDefinitionReaderService();
    this.setState({
      attributeDefinition: attributeDefinitionReaderService.readFile
    })
  }

  render() {
    const BlankPage = () => <div className="blank-page"/>;
    return (
      <Router>
        <div>
          <nav className="navbar navbar-expand-sm">
            <div className="navbar-nav d-flex flex-row">
              <Link className="nav-item nav-link mx-2" to={AttributeList.LOADFILE}>Load</Link>
              <button className="nav-item btn btn-primary mx-2" onClick={ (e) => this.onExport(e) }>Export</button>
            </div>
          </nav>
          <div className="container">
            <div className="row">
              <div className={ ' left-container col-3 d-flex' }>
                <Route path={ AttributeList.HOME }
                       render={ routeProps => this.renderAnimalList(routeProps) }/>
              </div>
              <div className={ ' right-container col-9 d-flex' }>
                <Switch>
                  <Route exact path={ AttributeList.HOME } component={ Greeting }/>
                  <Route exact path={ AttributeList.LOADFILE }
                         render={ routeProps => this.renderAnimalDefinitionLoader(routeProps) }/>
                  <Route exact path={ AttributeList.NEW }
                         render={ routeProps => this.renderAttributeGroupsForNewAnimal(routeProps) }/>
                  <Route path="/animal/:id"
                         render={ routeProps => this.renderAttributeGroupsForThisAnimal(routeProps) }/>
                  <Route component={ BlankPage }/>
                </Switch>
              </div>
            </div>
          </div>
        </div>
      </Router>
    )
  }
}

AttributeList.propTypes = {
  onExport: PropTypes.func.isRequired
};

export default AttributeList