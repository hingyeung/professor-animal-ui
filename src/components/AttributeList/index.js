import AttributeList from './AttributeList.jsx';

export default AttributeList;
