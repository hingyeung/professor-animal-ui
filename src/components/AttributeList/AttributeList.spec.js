import React from 'react';
import AttributeList from './AttributeList';
import {MemoryRouter} from 'react-router';
import AnimalDefinitionLoader from 'components/AnimalDefinitionLoader';
import {mount} from 'enzyme';
import Greeting from "../Greeting";
import LoadableAnimalForm from "components/Loadables/LoadableAnimalForm";
import LoadableAnimalList from "components/Loadables/LoadableAnimalList";

jest.mock('components/Loadables/LoadableAnimalList');
jest.mock('components/Loadables/LoadableAnimalForm');

describe('AttributeList', function () {
  const MOCK_ON_EXPORT = jest.fn();
  const animalModel = [{ id: "animalId", name: "Lion", physical: ["legs", "tail"] }];
  const animalFile = JSON.stringify(animalModel);

  beforeEach(function () {
    MOCK_ON_EXPORT.mockClear();
  });

  it('should render home route', function () {
    // https://github.com/ReactTraining/react-router/issues/5579#issuecomment-333401692
    const wrapper = mount(
      <MemoryRouter initialEntries={ [{pathname: "/", key: 'testKey'}] }>
        <AttributeList onExport={ MOCK_ON_EXPORT }/>
      </MemoryRouter>
    );

    expect(wrapper.find(Greeting)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  it('should render load-page route', function () {
    const wrapper = mount(
      <MemoryRouter initialEntries={ [{pathname: "/load", key: 'testKey', state: {activeAnimalId: "id"}}] }>
        <AttributeList onExport={ MOCK_ON_EXPORT }/>
      </MemoryRouter>
    );

    expect(wrapper.find(AnimalDefinitionLoader)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  it('should render animal route with greeting when animal definition is not loaded', function () {
    const wrapper = mount(
      <MemoryRouter initialEntries={ [{pathname: "/animal/animalId", key: 'testKey', state: {activeAnimalId: "animalId"}}] }>
        <AttributeList onExport={ MOCK_ON_EXPORT }/>
      </MemoryRouter>
    );

    expect(wrapper.find(Greeting)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();

    wrapper.unmount();
  });

  it('should render animal route with animal form when animal definition is loaded', function () {
    const wrapper = mount(
      <MemoryRouter initialEntries={ [{pathname: "/animal/animalId", key: 'testKey'}] }>
        <AttributeList onExport={ MOCK_ON_EXPORT }/>
      </MemoryRouter>
    );
    const component = wrapper.find('AttributeList');

    component.instance().onAnimalDefinitionLoaded(animalFile);
    wrapper.update();

    expect(wrapper.find(LoadableAnimalList)).toHaveLength(1);
    expect(wrapper.find(LoadableAnimalForm)).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();

    // setTimeout(() => validateDOM(wrapper), 1000);
    wrapper.unmount();
  });

  it('should call onExport() in props with animalDefinition in current state when export button is clicked', function () {
    const wrapper = mount(
      <MemoryRouter initialEntries={ [{pathname: "/animal/animalId", key: 'testKey'}] }>
        <AttributeList onExport={ MOCK_ON_EXPORT }/>
      </MemoryRouter>
    );
    const component = wrapper.find('AttributeList').instance();

    component.onAnimalDefinitionLoaded(animalFile);
    wrapper.find('button#export-btn').simulate('click');
    expect(MOCK_ON_EXPORT).toHaveBeenCalledTimes(1);
    // https://stackoverflow.com/a/47318973
    expect(MOCK_ON_EXPORT).toHaveBeenCalledWith(component.state.animalDefinition);
  });

  it('should show blank form', function () {
    const wrapper = mount(
      <MemoryRouter initialEntries={[{pathname: "/new", key: "testKey"}]}>
        <AttributeList onExport={MOCK_ON_EXPORT}/>
      </MemoryRouter>
    );

    expect(wrapper.find(LoadableAnimalList)).toHaveLength(1);
    expect(wrapper.find(LoadableAnimalForm)).toHaveLength(1);
  });

  it('should call onExport function in props when onExport is called', function () {
    const wrapper = mount(
      <MemoryRouter initialEntries={[{pathname: "/new", key: "testKey"}]}>
        <AttributeList onExport={MOCK_ON_EXPORT}/>
      </MemoryRouter>
    );

    wrapper.find(AttributeList).instance().onExport({preventDefault: jest.fn()});
    expect(MOCK_ON_EXPORT).toHaveBeenCalledTimes(1);
  });

  it('should update state with the current animal when onFormSubmit is called', function () {
    const wrapper = mount(
      <MemoryRouter initialEntries={[{pathname: "/new", key: "testKey"}]}>
        <AttributeList onExport={MOCK_ON_EXPORT}/>
      </MemoryRouter>
    );

    const mockRouteHistory = {push: jest.fn()};
    const attributeListInstance = wrapper.find(AttributeList).instance();
    const mockAnimal = {id: "id"};
    attributeListInstance.onFormSubmit(mockAnimal, mockRouteHistory);
    expect(attributeListInstance.state.animalDefinition).toEqual({id: mockAnimal});
  })
});