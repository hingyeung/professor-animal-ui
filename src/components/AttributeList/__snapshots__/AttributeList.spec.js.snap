// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttributeList should render animal route with animal form when animal definition is loaded 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/animal/animalId",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/animal/animalId",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/animal/animalId",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AttributeList
      onExport={[Function]}
    >
      <Component>
        <div>
          <div>
            <nav
              className="navbar navbar-expand-sm"
            >
              <div
                className="navbar-nav d-flex flex-row"
              >
                <Link
                  className="nav-item nav-link mx-2"
                  replace={false}
                  to="/load"
                >
                  <a
                    className="nav-item nav-link mx-2"
                    href="/load"
                    onClick={[Function]}
                  >
                    Load
                  </a>
                </Link>
                <button
                  className="nav-item btn btn-primary mx-2"
                  id="export-btn"
                  onClick={[Function]}
                >
                  Export
                </button>
              </div>
            </nav>
            <div
              className="container"
            >
              <div
                className="row"
              >
                <div
                  className=" left-container col-3 d-flex"
                >
                  <Route
                    path="/"
                    render={[Function]}
                  >
                    <LoadableAnimalList
                      animals={
                        Object {
                          "animalId": Object {
                            "attributeMap": Object {
                              "behaviours": Object {
                                "bark": false,
                                "climb tree": false,
                                "eat both plants and animals": false,
                                "fly": false,
                                "hibernate in winter": false,
                                "hunt in a group": false,
                                "lay eggs": false,
                                "live in a group": false,
                                "live in a large group": false,
                                "live in burrows": false,
                                "meow": false,
                                "only eat meat": false,
                                "only eat plants": false,
                                "only live in Australia": false,
                                "roar": false,
                                "slither across flat surfaces without legs": false,
                                "sting": false,
                              },
                              "considerations": Object {
                                "as dangerous to humans": false,
                                "as helpful to humans": false,
                                "as most dangerous animals in Africa": false,
                                "to be friendly to humans": false,
                                "to be very intelligent": false,
                              },
                              "diet": Object {
                                "algae": false,
                                "bird": false,
                                "bone": false,
                                "fish": false,
                                "grass": false,
                                "grass and plants": false,
                                "insects": false,
                                "jellyfish": false,
                                "krill": false,
                                "nectar from flowers": false,
                                "pet food": false,
                                "plankton": false,
                                "sea grass and seaweed": false,
                                "seals": false,
                                "tiny fish and fish eggs": false,
                              },
                              "physical": Object {
                                "a hard shell": false,
                                "a pouch where their newborns live": false,
                                "black stripes on its body": false,
                                "dark hair": false,
                                "fins": false,
                                "hooves": false,
                                "huge mouth": false,
                                "large animal": false,
                                "large ears": false,
                                "large jaw": false,
                                "legs": true,
                                "long arms": false,
                                "long body": false,
                                "long legs": false,
                                "long neck": false,
                                "long tongue": false,
                                "long trunk": false,
                                "mane": false,
                                "many sharp teeth": false,
                                "no teeth": false,
                                "pouch on its tummy": false,
                                "scales": false,
                                "sharp claws": false,
                                "short legs": false,
                                "six legs": false,
                                "strong talon": false,
                                "tail": true,
                                "tentacles": false,
                                "thick, armour-like skin": false,
                                "thumbs like humans": false,
                                "tusks": false,
                                "two antenna": false,
                                "wings": false,
                              },
                              "possible_behaviours": Object {
                                "be poisonous": false,
                                "domesticated": false,
                                "enlarge its body by swallowing huge amount of water to scare off predators": false,
                                "jump out of the water": false,
                                "jump very high": false,
                                "live in freshwater as well as salt water": false,
                                "swim": false,
                                "use tools": false,
                              },
                              "types": Object {
                                "a highly intelligent animal": false,
                                "ape": false,
                                "big cat": false,
                                "bird": false,
                                "bird of prey": false,
                                "cold-blooded animal": false,
                                "fish": false,
                                "insect": false,
                                "invertebrate": false,
                                "large bird": false,
                                "mammal": false,
                                "marsupial": false,
                                "nocturnal animal": false,
                                "pet": false,
                                "predator": false,
                                "reptile": false,
                                "small animal": false,
                                "social animal": false,
                                "solitary animal": false,
                                "territorial animal": false,
                                "vertebrate": false,
                              },
                            },
                            "id": "animalId",
                            "name": "Lion",
                          },
                        }
                      }
                    >
                      <div>
                        LoadableAnimalList
                      </div>
                    </LoadableAnimalList>
                  </Route>
                </div>
                <div
                  className=" right-container col-9 d-flex"
                >
                  <Switch>
                    <Route
                      computedMatch={
                        Object {
                          "isExact": true,
                          "params": Object {
                            "id": "animalId",
                          },
                          "path": "/animal/:id",
                          "url": "/animal/animalId",
                        }
                      }
                      location={
                        Object {
                          "hash": "",
                          "key": "testKey",
                          "pathname": "/animal/animalId",
                          "search": "",
                        }
                      }
                      path="/animal/:id"
                      render={[Function]}
                    >
                      <LoadableAnimalForm
                        animal={
                          Object {
                            "attributeMap": Object {
                              "behaviours": Object {
                                "bark": false,
                                "climb tree": false,
                                "eat both plants and animals": false,
                                "fly": false,
                                "hibernate in winter": false,
                                "hunt in a group": false,
                                "lay eggs": false,
                                "live in a group": false,
                                "live in a large group": false,
                                "live in burrows": false,
                                "meow": false,
                                "only eat meat": false,
                                "only eat plants": false,
                                "only live in Australia": false,
                                "roar": false,
                                "slither across flat surfaces without legs": false,
                                "sting": false,
                              },
                              "considerations": Object {
                                "as dangerous to humans": false,
                                "as helpful to humans": false,
                                "as most dangerous animals in Africa": false,
                                "to be friendly to humans": false,
                                "to be very intelligent": false,
                              },
                              "diet": Object {
                                "algae": false,
                                "bird": false,
                                "bone": false,
                                "fish": false,
                                "grass": false,
                                "grass and plants": false,
                                "insects": false,
                                "jellyfish": false,
                                "krill": false,
                                "nectar from flowers": false,
                                "pet food": false,
                                "plankton": false,
                                "sea grass and seaweed": false,
                                "seals": false,
                                "tiny fish and fish eggs": false,
                              },
                              "physical": Object {
                                "a hard shell": false,
                                "a pouch where their newborns live": false,
                                "black stripes on its body": false,
                                "dark hair": false,
                                "fins": false,
                                "hooves": false,
                                "huge mouth": false,
                                "large animal": false,
                                "large ears": false,
                                "large jaw": false,
                                "legs": true,
                                "long arms": false,
                                "long body": false,
                                "long legs": false,
                                "long neck": false,
                                "long tongue": false,
                                "long trunk": false,
                                "mane": false,
                                "many sharp teeth": false,
                                "no teeth": false,
                                "pouch on its tummy": false,
                                "scales": false,
                                "sharp claws": false,
                                "short legs": false,
                                "six legs": false,
                                "strong talon": false,
                                "tail": true,
                                "tentacles": false,
                                "thick, armour-like skin": false,
                                "thumbs like humans": false,
                                "tusks": false,
                                "two antenna": false,
                                "wings": false,
                              },
                              "possible_behaviours": Object {
                                "be poisonous": false,
                                "domesticated": false,
                                "enlarge its body by swallowing huge amount of water to scare off predators": false,
                                "jump out of the water": false,
                                "jump very high": false,
                                "live in freshwater as well as salt water": false,
                                "swim": false,
                                "use tools": false,
                              },
                              "types": Object {
                                "a highly intelligent animal": false,
                                "ape": false,
                                "big cat": false,
                                "bird": false,
                                "bird of prey": false,
                                "cold-blooded animal": false,
                                "fish": false,
                                "insect": false,
                                "invertebrate": false,
                                "large bird": false,
                                "mammal": false,
                                "marsupial": false,
                                "nocturnal animal": false,
                                "pet": false,
                                "predator": false,
                                "reptile": false,
                                "small animal": false,
                                "social animal": false,
                                "solitary animal": false,
                                "territorial animal": false,
                                "vertebrate": false,
                              },
                            },
                            "id": "animalId",
                            "name": "Lion",
                          }
                        }
                        onFormSubmit={[Function]}
                        routeHistory={
                          Object {
                            "action": "POP",
                            "block": [Function],
                            "canGo": [Function],
                            "createHref": [Function],
                            "entries": Array [
                              Object {
                                "hash": "",
                                "key": "testKey",
                                "pathname": "/animal/animalId",
                                "search": "",
                              },
                            ],
                            "go": [Function],
                            "goBack": [Function],
                            "goForward": [Function],
                            "index": 0,
                            "length": 1,
                            "listen": [Function],
                            "location": Object {
                              "hash": "",
                              "key": "testKey",
                              "pathname": "/animal/animalId",
                              "search": "",
                            },
                            "push": [Function],
                            "replace": [Function],
                          }
                        }
                      >
                        <div>
                          LoadableAnimalForm
                        </div>
                      </LoadableAnimalForm>
                    </Route>
                  </Switch>
                </div>
              </div>
            </div>
          </div>
        </div>
      </Component>
    </AttributeList>
  </Router>
</MemoryRouter>
`;

exports[`AttributeList should render animal route with greeting when animal definition is not loaded 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/animal/animalId",
        "state": Object {
          "activeAnimalId": "animalId",
        },
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/animal/animalId",
            "search": "",
            "state": Object {
              "activeAnimalId": "animalId",
            },
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/animal/animalId",
          "search": "",
          "state": Object {
            "activeAnimalId": "animalId",
          },
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AttributeList
      onExport={[Function]}
    >
      <Component>
        <div>
          <div>
            <nav
              className="navbar navbar-expand-sm"
            >
              <div
                className="navbar-nav d-flex flex-row"
              >
                <Link
                  className="nav-item nav-link mx-2"
                  replace={false}
                  to="/load"
                >
                  <a
                    className="nav-item nav-link mx-2"
                    href="/load"
                    onClick={[Function]}
                  >
                    Load
                  </a>
                </Link>
                <button
                  className="nav-item btn btn-primary mx-2"
                  id="export-btn"
                  onClick={[Function]}
                >
                  Export
                </button>
              </div>
            </nav>
            <div
              className="container"
            >
              <div
                className="row"
              >
                <div
                  className=" left-container col-3 d-flex"
                >
                  <Route
                    path="/"
                    render={[Function]}
                  >
                    <LoadableAnimalList
                      activeAnimalId="animalId"
                      animals={Object {}}
                    >
                      <div>
                        LoadableAnimalList
                      </div>
                    </LoadableAnimalList>
                  </Route>
                </div>
                <div
                  className=" right-container col-9 d-flex"
                >
                  <Switch>
                    <Route
                      computedMatch={
                        Object {
                          "isExact": true,
                          "params": Object {
                            "id": "animalId",
                          },
                          "path": "/animal/:id",
                          "url": "/animal/animalId",
                        }
                      }
                      location={
                        Object {
                          "hash": "",
                          "key": "testKey",
                          "pathname": "/animal/animalId",
                          "search": "",
                          "state": Object {
                            "activeAnimalId": "animalId",
                          },
                        }
                      }
                      path="/animal/:id"
                      render={[Function]}
                    >
                      <Greeting>
                        <div
                          className="greeting-container"
                        >
                          <h1>
                            Click Load to load an existing animal definition.
                          </h1>
                        </div>
                      </Greeting>
                    </Route>
                  </Switch>
                </div>
              </div>
            </div>
          </div>
        </div>
      </Component>
    </AttributeList>
  </Router>
</MemoryRouter>
`;

exports[`AttributeList should render home route 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AttributeList
      onExport={[Function]}
    >
      <Component>
        <div>
          <div>
            <nav
              className="navbar navbar-expand-sm"
            >
              <div
                className="navbar-nav d-flex flex-row"
              >
                <Link
                  className="nav-item nav-link mx-2"
                  replace={false}
                  to="/load"
                >
                  <a
                    className="nav-item nav-link mx-2"
                    href="/load"
                    onClick={[Function]}
                  >
                    Load
                  </a>
                </Link>
                <button
                  className="nav-item btn btn-primary mx-2"
                  id="export-btn"
                  onClick={[Function]}
                >
                  Export
                </button>
              </div>
            </nav>
            <div
              className="container"
            >
              <div
                className="row"
              >
                <div
                  className=" left-container col-3 d-flex"
                >
                  <Route
                    path="/"
                    render={[Function]}
                  >
                    <LoadableAnimalList
                      animals={Object {}}
                    >
                      <div>
                        LoadableAnimalList
                      </div>
                    </LoadableAnimalList>
                  </Route>
                </div>
                <div
                  className=" right-container col-9 d-flex"
                >
                  <Switch>
                    <Route
                      component={[Function]}
                      computedMatch={
                        Object {
                          "isExact": true,
                          "params": Object {},
                          "path": "/",
                          "url": "/",
                        }
                      }
                      exact={true}
                      location={
                        Object {
                          "hash": "",
                          "key": "testKey",
                          "pathname": "/",
                          "search": "",
                        }
                      }
                      path="/"
                    >
                      <Greeting
                        history={
                          Object {
                            "action": "POP",
                            "block": [Function],
                            "canGo": [Function],
                            "createHref": [Function],
                            "entries": Array [
                              Object {
                                "hash": "",
                                "key": "testKey",
                                "pathname": "/",
                                "search": "",
                              },
                            ],
                            "go": [Function],
                            "goBack": [Function],
                            "goForward": [Function],
                            "index": 0,
                            "length": 1,
                            "listen": [Function],
                            "location": Object {
                              "hash": "",
                              "key": "testKey",
                              "pathname": "/",
                              "search": "",
                            },
                            "push": [Function],
                            "replace": [Function],
                          }
                        }
                        location={
                          Object {
                            "hash": "",
                            "key": "testKey",
                            "pathname": "/",
                            "search": "",
                          }
                        }
                        match={
                          Object {
                            "isExact": true,
                            "params": Object {},
                            "path": "/",
                            "url": "/",
                          }
                        }
                      >
                        <div
                          className="greeting-container"
                        >
                          <h1>
                            Click Load to load an existing animal definition.
                          </h1>
                        </div>
                      </Greeting>
                    </Route>
                  </Switch>
                </div>
              </div>
            </div>
          </div>
        </div>
      </Component>
    </AttributeList>
  </Router>
</MemoryRouter>
`;

exports[`AttributeList should render load-page route 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
        "pathname": "/load",
        "state": Object {
          "activeAnimalId": "id",
        },
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/load",
            "search": "",
            "state": Object {
              "activeAnimalId": "id",
            },
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/load",
          "search": "",
          "state": Object {
            "activeAnimalId": "id",
          },
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <AttributeList
      onExport={[Function]}
    >
      <Component>
        <div>
          <div>
            <nav
              className="navbar navbar-expand-sm"
            >
              <div
                className="navbar-nav d-flex flex-row"
              >
                <Link
                  className="nav-item nav-link mx-2"
                  replace={false}
                  to="/load"
                >
                  <a
                    className="nav-item nav-link mx-2"
                    href="/load"
                    onClick={[Function]}
                  >
                    Load
                  </a>
                </Link>
                <button
                  className="nav-item btn btn-primary mx-2"
                  id="export-btn"
                  onClick={[Function]}
                >
                  Export
                </button>
              </div>
            </nav>
            <div
              className="container"
            >
              <div
                className="row"
              >
                <div
                  className=" left-container col-3 d-flex"
                >
                  <Route
                    path="/"
                    render={[Function]}
                  >
                    <LoadableAnimalList
                      activeAnimalId="id"
                      animals={Object {}}
                    >
                      <div>
                        LoadableAnimalList
                      </div>
                    </LoadableAnimalList>
                  </Route>
                </div>
                <div
                  className=" right-container col-9 d-flex"
                >
                  <Switch>
                    <Route
                      computedMatch={
                        Object {
                          "isExact": true,
                          "params": Object {},
                          "path": "/load",
                          "url": "/load",
                        }
                      }
                      exact={true}
                      location={
                        Object {
                          "hash": "",
                          "key": "testKey",
                          "pathname": "/load",
                          "search": "",
                          "state": Object {
                            "activeAnimalId": "id",
                          },
                        }
                      }
                      path="/load"
                      render={[Function]}
                    >
                      <AnimalDefinitionLoader
                        onFileLoaded={[Function]}
                      >
                        <div>
                          <form>
                            <input
                              className="form-control"
                              onChange={[Function]}
                              type="file"
                            />
                          </form>
                        </div>
                      </AnimalDefinitionLoader>
                    </Route>
                  </Switch>
                </div>
              </div>
            </div>
          </div>
        </div>
      </Component>
    </AttributeList>
  </Router>
</MemoryRouter>
`;
